File: core/src/main/java/de/qabel/core/util/LazyWeakHashMap.java
package de.qabel.core.util;

import java.util.WeakHashMap;

public class LazyWeakHashMap<K, V> extends WeakHashMap<K, V> implements LazyMap<K, V> {
    @Override
    public V getOrDefault(K key, CheckedFunction<K, V> defaultValueFactory) {
        synchronized (this) {
            if (!containsKey(key)) {
                try {
                    put(key, defaultValueFactory.apply(key));
                } catch (Exception e) {
                    throw new RuntimeException(e.getMessage(), e);
                }
            }
            return get(key);
        }
    }
}

File: core/src/main/java/de/qabel/core/util/LazyWeakHashMap.kt
package de.qabel.core.util

import java.util.WeakHashMap

class LazyWeakHashMap<K, V> : WeakHashMap<K, V>(), LazyMap<K, V> {
    override fun getOrDefault(key: K, defaultValueFactory: CheckedFunction<K, V>): V {
        synchronized(this) {
            if (!containsKey(key)) {
                try {
                    put(key, defaultValueFactory.apply(key))
                } catch (e: Exception) {
                    throw RuntimeException(e.message, e)
                }

            }
            return get(key)
        }
    }
}
